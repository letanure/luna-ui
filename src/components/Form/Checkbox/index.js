import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

import FontAwesomeIcon from '@fortawesome/react-fontawesome';

/**
 * Checkbox component.
 * E.g.:
 * ```html
 *  <Form.Checkbox id="ex1" kind="primary" label="pop" group="ex" value="1" />
 *  <Form.Checkbox id="ex2" kind="default" label="rock" group="ex" value="2" />
 *  <Form.Checkbox id="ex3" kind="success" label="funk" group="ex" value="3" />
 *  <Form.Checkbox id="ex4" kind="danger" label="samba" group="ex" value="4" />
 *  <Form.Checkbox id="ex5" kind="warning" label="hiphop" group="ex" value="5" />
 *  <Form.Checkbox id="ex6" kind="danger" label="pagode" group="ex" value="6" config={{ disabled: true }} />
 * ```
 */
class CheckboxComponent extends React.PureComponent {
  render() {
    const { config, kind, value, id, label, group, className } = this.props;

    const Input = styled.input`
      visibility: hidden;
      &:checked + label {
        background-color: ${(props) => props.theme.form.checkbox.types[kind]};
      }

      &::not(:checked) + label {
        background-color: ${(props) => props.theme.form.checkbox.checkIcon};
      }

      &:disabled ~ label {
        cursor: not-allowed;
        color: ${(props) => props.theme.form.checkbox.disabledText};
      }

      &:disabled + label {
        color: ${(props) => props.theme.form.checkbox.disabled};
        background-color: ${(props) => props.theme.form.checkbox.disabled};
        border-color: ${(props) => props.theme.form.checkbox.disabledBorder};
      }
    `;

    const Checkbox = styled.label`
      width: 16px;
      height: 16px;
      padding: 2px;
      cursor: pointer;
      font-size: 10px;
      border-radius: 2px;
      position: absolute;
      left: 11px;
      transition: background-color .15s linear;
      color: ${(props) => props.theme.form.checkbox.checkIcon};
      background-color: ${(props) => props.theme.form.checkbox.checkIcon};
      border: 1px solid ${(props) => props.theme.form.checkbox.types[kind]};
      & + label {
        position: absolute;
        left: 27px;
        cursor: pointer;
      }
    `;

    const CheckboxWrapper = styled.div`
      height: 40px;
      padding: 11px;
      display: flex;
      position: relative;
    `;

    return (
      <CheckboxWrapper className={className}>
        <Input readOnly id={id} name={group} type="checkbox" {...config} value={value} />
        <Checkbox htmlFor={id}>
          <FontAwesomeIcon icon="check" />
        </Checkbox>
        <label htmlFor={id}>{label}</label>
      </CheckboxWrapper>
    );
  }
}

CheckboxComponent.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  group: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  kind: PropTypes.string.isRequired,
  config: PropTypes.object
};

CheckboxComponent.defaultProps = {
  className: '',
  label: '',
  kind: 'primary',
  config: {}
};

export default CheckboxComponent;
