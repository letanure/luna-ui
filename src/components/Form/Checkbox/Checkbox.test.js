import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../../config/defaultTheme';
import ThemeProvider from '../../Theme/index';

import Checkbox from './index.js';

describe('<Checkbox />', () => {
  let checkboxList;
  let onChange;

  beforeEach(() => {
    onChange = sinon.spy();

    checkboxList =
      <ThemeProvider theme={theme}>
        <Checkbox id="ex1" kind="primary" label="pop" group="ex" value="1" />
        <Checkbox id="ex2" kind="default" label="rock" group="ex" value="2" config={{ checked: true, onChange }} />
        <Checkbox id="ex3" kind="default" label="rock" group="ex" value="2" config={{ disabled: true }} />
      </ThemeProvider>;
  });

  test('To be checked', () => {
    const wrapper = mount(checkboxList);
    expect(wrapper.find('input').at(1)
      .getDOMNode().checked).toBeTruthy();
  });

  test('Call onChange', () => {
    const wrapper = mount(checkboxList);
    const checkbox = () => wrapper.find('input').at(1);
    checkbox().simulate('change');
    expect(onChange).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot: List of Checkboxes', () => {
    const tree = renderer
      .create(checkboxList)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
