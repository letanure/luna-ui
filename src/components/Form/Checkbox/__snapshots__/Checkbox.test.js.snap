// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> toMatchSnapshot: List of Checkboxes 1`] = `
<div>
  <div
    className="sc-kgoBCf ivXsRw"
  >
    <input
      className="sc-kAzzGY bpDdd"
      id="ex1"
      name="ex"
      readOnly={true}
      type="checkbox"
      value="1"
    />
    <label
      className="sc-chPdSV fwYjhg"
      htmlFor="ex1"
    />
    <label
      htmlFor="ex1"
    >
      pop
    </label>
  </div>
  <div
    className="sc-dxgOiQ lndBkL"
  >
    <input
      checked={true}
      className="sc-kGXeez eYqwZQ"
      id="ex2"
      name="ex"
      onChange={[Function]}
      readOnly={true}
      type="checkbox"
      value="2"
    />
    <label
      className="sc-kpOJdX cnkErW"
      htmlFor="ex2"
    />
    <label
      htmlFor="ex2"
    >
      rock
    </label>
  </div>
  <div
    className="sc-eNQAEJ bdadmr"
  >
    <input
      className="sc-ckVGcZ jhrodw"
      disabled={true}
      id="ex3"
      name="ex"
      readOnly={true}
      type="checkbox"
      value="2"
    />
    <label
      className="sc-jKJlTe Zsqdt"
      htmlFor="ex3"
    />
    <label
      htmlFor="ex3"
    >
      rock
    </label>
  </div>
</div>
`;
